[CI] Quantization workflow

SunMarc opened the pull request:
# What does this PR do ? 
This PR adds a workflow for quantization tests + related dockerfile. Since we merged the [HfQuantizer PR](https://github.com/huggingface/transformers/pull/26610), the community started integrating their own quantizers into transformers (e.g. [AQML](https://github.com/huggingface/transformers/pull/28928) and much more in the future). This will lead to many third party libraries in the Dockerfile `huggingface/transformers-all-latest-gpu`. To limit the impact of these libraries on transformers tests, I propose to create a seperate dockerfile + workflow.

SunMarc committed:
[CI] Quantization workflow

SunMarc committed:
build dockerfile

SunMarc committed:
fix dockerfile

SunMarc committed:
update self-cheduled.yml

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29046). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

younesbelkada reviewed (commented):
Thank you very much ! Can you try to trigger a run of the quantization slow tests and make sure we get the slack notifications (after building the quantization docker image) ? 🙏 
For that you can comment out all jobs except the quantization tests job in `self-scheduled.yml` & `build-docker.yml` and modify the run condition on on push to the name of the branch you are working on right now (there might be a better way to trigger only the quantization tests but all tests are on the same workflow file)

younesbelkada commented on .github/workflows/self-scheduled.yml:
```diff
@@ -297,6 +297,57 @@ jobs:
+      options: --gpus all --shm-size "16gb" --ipc host -v /mnt/cache/.cache/huggingface:/mnt/cache/
+    needs: setup
+    steps:
+      - name: Update clone
+        working-directory: /transformers
+        run: git fetch && git checkout ${{ github.sha }}
+
+      - name: Reinstall transformers in edit mode (remove the one installed during docker image build)
+        working-directory: /transformers
+        run: python3 -m pip uninstall -y transformers && python3 -m pip install -e .
```
```suggestion
      - uses: actions/checkout@v3
```
I think this can be completely removed in favor of uses: actions/checkout@v3: https://github.com/actions/checkout

See e.g.: https://github.com/huggingface/trl/blob/main/.github/workflows/slow-tests.yml#L33C15-L33C31

SunMarc replied:
Thank you, I will first test it without and try to add that when I'm able to run the workflow successfully !

younesbelkada replied:
You also need to call `pip install -e .` at the begining of the job after the actions/checkout

SunMarc replied:
I will leave it as it is since I think that @ydshieh prefer not to use `actions/checkout@v3`. We can still make the change in this [PR](https://github.com/huggingface/transformers/pull/29197) if we see that it works !

ydshieh replied:
yes, I talked to @younesbelkada . Il fait n'importe de quoi 😆  (joking)

younesbelkada replied:
hahaha

younesbelkada reviewed (commented):
I left a comment ! (this can be applied to all our workflow file that used that logic for checking out to the current main branch)

SunMarc committed:
test build dockerfile on push

SunMarc commented:
> For that you can comment out all jobs except the quantization tests job in self-scheduled.yml & build-docker.yml and modify the run condition on on push to the name of the branch you are working on right now (there might be a better way to trigger only the quantization tests but all tests are on the same workflow file)

I tried to build the docker image but it gives the following error: 

`ERROR: Could not install packages due to an OSError: [Errno 28] No space left on device`.
Do you know how to fix that ? 

Also, I see that  `transformers-all-latest-gpu` docker image is not being updated for the last two days since the [installation](https://github.com/huggingface/transformers/actions/runs/7924158967/job/21635225922) fails because of aqml library that requires python 3.10 at least and we uses 3.8 for now. We will have to change that in the quantization dockerfile.

SunMarc committed:
fix torch install

SunMarc committed:
udapte to python 3.10

BlackSamorez commented:
The only reason `aqlm` requires `python>=3.10` is a single `match-case` statement in a non-critical place.

I was able to run `aqlm` on `python 3.8` no problem. I replace the statement with an `if-else` statement and lower the requirement if necessary.

younesbelkada commented:
@SunMarc thanks! 
in trl i build a docker image with python 3.10: https://github.com/huggingface/trl/blob/main/docker/trl-source-gpu/Dockerfile maybe you can take some inspiration from that dockerfile? 🙏 I am not sure why you are getting that error currently as the commands looks correct. 
@BlackSamorez yes that would be great if you can also support python 3.8 for AQLM 🙏 Thanks !

SunMarc commented:
> I was able to run aqlm on python 3.8 no problem otherwise. I can replace the statement with an if-else statement and lower the requirement if necessary.

Yes that would be for the best ! I prefer to keep running the quantization tests with python 3.8 since this is what we are actually doing for all transformers tests ! Moreover, it will be better for the users since we keep the requirement low. LMK when it is done ! @BlackSamorez 
Otherwise, I will modify the dockerfile and use conda to install python 3.10 as suggested by @younesbelkada.

BlackSamorez commented:
@SunMarc `aqlm` will support python `>=3.8` starting version `1.0.2`. I'm [1 PR](https://github.com/Vahe1994/AQLM/pull/26) away from releasing it.

SunMarc commented:
Perfect ! I will wait for your PR to be merged then + release then if it doesn't take too much time. Please keep me updated ! Otherwise, I can first merge this PR without aqlm and add it afterwards.

BlackSamorez commented:
@SunMarc `aqlm==1.0.2` is out. May I ask you to please update the docker images?

SunMarc committed:
update aqlm version

SunMarc commented:
I was able to build the image but I don't have the permission to push the image cc @ydshieh 

`#22 ERROR: failed to push huggingface/transformers-quantization-latest-gpu: push access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed`

ydshieh commented:
@SunMarc  Do you builded/pushed via transformers' github actions? If so, do you have a job run link?

SunMarc commented:
Yes ! Here's the link to the [job] (https://github.com/huggingface/transformers/actions/runs/7979164961)

younesbelkada commented:
it's quite strange because the workflow was indeed able to login: https://github.com/huggingface/transformers/actions/runs/7979164961/job/21802381289#step:5:1 but fails to push ...

younesbelkada commented:
In TRL and PEFT I can confirm the build & push works fine :  https://github.com/huggingface/peft/actions/workflows/build_docker_images.yml / https://github.com/huggingface/trl/actions/workflows/docker-build.yml so it's not a token issue as we use the same except if the token has expired for transformers cc @glegendre01

ydshieh commented:
Hi @SunMarc it's because there is some change in the infra team on docker hub stuff.

The repository `huggingface/transformers-quantization-latest-gpu` has to be created on docker Hub first, then only after this, you can push to it.

I will ask for it.

ydshieh commented:
BTW, I will review this PR tomorrow or Friday 🙏

SunMarc committed:
uncomment build dockerfile

SunMarc committed:
tests if the scheduler works

SunMarc committed:
fix docker

SunMarc committed:
do not trigger on psuh again

SunMarc committed:
add additional runs

SunMarc committed:
test again

SunMarc committed:
all good

SunMarc committed:
Merge remote-tracking branch 'upstream/main' into add-quantization-workflow

SunMarc committed:
style

younesbelkada commented on .github/workflows/self-scheduled.yml:
```diff
@@ -297,6 +297,57 @@ jobs:
+      - name: Run quantization tests on GPU
+        working-directory: /transformers
+        run: |
+          pip install -r examples/pytorch/_tests_requirements.txt
```
```suggestion
```

SunMarc committed:
Update .github/workflows/self-scheduled.yml

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

ydshieh commented on docker/transformers-all-latest-gpu/Dockerfile:
```diff
@@ -46,22 +46,10 @@ RUN python3 -m pip install --no-cache-dir git+https://github.com/huggingface/acc
-
-# For bettertransformer + gptq
+# For bettertransformer
 RUN python3 -m pip install --no-cache-dir git+https://github.com/huggingface/optimum@main#egg=optimum
```
Do we count this as `non quantization` tests?

younesbelkada replied:
I think we need to keep that as optimum is a hard requirement to use GPTQ

ydshieh commented on docker/transformers-quantization-latest-gpu/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+# The following `ARG` are mainly used to specify the versions explicitly & directly in this docker file, and not meant
+# to be used as arguments for docker build (so far).
+
+ARG PYTORCH='2.1.1'
```
Let's aim to use `torch 2.2`. See #29208

ydshieh commented on docker/transformers-quantization-latest-gpu/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+ARG REF=main
+RUN git clone https://github.com/huggingface/transformers && cd transformers && git checkout $REF
+
+RUN [ ${#PYTORCH} -gt 0 ] && VERSION='torch=='$PYTORCH'.*' ||  VERSION='torch'; echo "export VERSION='$VERSION'" >> ~/.profile
+RUN echo torch=$VERSION
+# `torchvision` and `torchaudio` should be installed along with `torch`, especially for nightly build.
+# Currently, let's just use their latest releases (when `torch` is installed with a release version)
+RUN python3 -m pip install --no-cache-dir -U $VERSION torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/$CUDA
```
make sure at the end, we do get `torch 2.2`. See my comment in the change in #29208. Otherwise, move this after `./transformers[dev]` step.

SunMarc replied:
Done ! I didn't have to move it after  `./transformers[dev]` step.

ydshieh commented on docker/transformers-quantization-latest-gpu/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+
+RUN python3 -m pip install --no-cache-dir -e ./transformers[dev]
+
+RUN python3 -m pip uninstall -y flax jax
```
we can also uninstall tensorflow. Or just use `./transformers[torch]` above (but we might need to install other stuffs manually, not sure)

(not a big deal, we can keep it if you are busy)

SunMarc replied:
I can switch to transformers[dev-torch] if this makes more sense.

ydshieh commented on .github/workflows/build-docker-images.yml:
```diff
@@ -307,3 +307,41 @@ jobs:
+    name: "Latest Pytorch + Quantization [dev]"
+     # Push CI doesn't need this image
+    if: inputs.image_postfix != '-push-ci'
+    runs-on: ubuntu-22.04
```
let's from now on move this to `[intel-cpu, 8-cpu, ci]` (our own VM, bigger RAM and disk)

SunMarc replied:
done !

ydshieh commented on .github/workflows/build-docker-images.yml:
```diff
@@ -307,3 +307,41 @@ jobs:
+    if: inputs.image_postfix != '-push-ci'
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Cleanup disk
+        run: |
+          sudo ls -l /usr/local/lib/
+          sudo ls -l /usr/share/
+          sudo du -sh /usr/local/lib/
+          sudo du -sh /usr/share/
+          sudo rm -rf /usr/local/lib/android
+          sudo rm -rf /usr/share/dotnet
+          sudo du -sh /usr/local/lib/
+          sudo du -sh /usr/share/
```
we can remove this now I guess (once we use `[intel-cpu, 8-cpu, ci]`)

SunMarc replied:
removed !

ydshieh commented on .github/workflows/build-docker-images.yml:
```diff
@@ -307,3 +307,41 @@ jobs:
+  latest-quantization-torch-docker:
+    name: "Latest Pytorch + Quantization [dev]"
+     # Push CI doesn't need this image
+    if: inputs.image_postfix != '-push-ci'
```
Just want to know if we intent to run `quantization` tests in a daily basis, or if we prefer to run them on each commit merged into `main`?

SunMarc replied:
I think that it is better to run them on a daily basis. There are many slows tests too + I think that the quantization tests are not easily broken by changes in transformers. The bigger issue is the third party libraries that might introduce breaking changes. cc @younesbelkada

younesbelkada replied:
Yes daily basis sounds great!

ydshieh replied:
OK

ydshieh commented on .github/workflows/self-scheduled.yml:
```diff
@@ -297,6 +297,56 @@ jobs:
           name: ${{ matrix.machine_type }}_run_tests_torch_cuda_extensions_gpu_test_reports
           path: /workspace/transformers/reports/${{ matrix.machine_type }}_tests_torch_cuda_extensions_gpu
 
+  run_tests_quantization_torch_gpu:
+    name: Quantization tests
```
I am not sure if we intend to keep this in this workflow file. Isn't the goal to move this outside to a separate workflow file ?

cc @younesbelkada

younesbelkada replied:
I think we can keep it there for now and I'll move it outside in my PR

ydshieh commented on utils/notification_service.py:
```diff
@@ -1043,6 +1043,7 @@ def prepare_reports(title, header, reports, to_truncate=True):
         "PyTorch pipelines": "run_tests_torch_pipeline_gpu",
         "TensorFlow pipelines": "run_tests_tf_pipeline_gpu",
         "Torch CUDA extension tests": "run_tests_torch_cuda_extensions_gpu_test_reports",
+        "Quantization tests": "run_tests_quantization_torch_gpu",
```
should be revised if we decide to move quantization tests outside this workflow file.

ydshieh reviewed (commented):
Thank  you @SunMarc (and @younesbelkada ) for this initiative.

I love this idea - quantization tests become more and more, and failing on them becomes hard to track in the whole daily CI.

I have some suggestions and questions for the current version though.

ydshieh commented:
@SunMarc I am going to merge #29208. Once done, there might be a conflict with this PR, but it should be easy to resolve. Ping me otherwise.

SunMarc committed:
Merge remote-tracking branch 'upstream/main' into add-quantization-workflow

SunMarc committed:
test build dockerfile with torch 2.2.0

SunMarc commented:
I've addressed all comments. I think that we are good to merge after I checked that the docker image is built correctly with torch 2.2.0 (currently facing some issues with login into dockerhub) cc @ydshieh

SunMarc committed:
fix extra

SunMarc committed:
clean

ArthurZucker commented on .github/workflows/self-scheduled.yml:
```diff
@@ -297,6 +297,56 @@ jobs:
+        working-directory: /transformers
+        run: git fetch && git checkout ${{ github.sha }}
+
+      - name: Reinstall transformers in edit mode (remove the one installed during docker image build)
+        working-directory: /transformers
+        run: python3 -m pip uninstall -y transformers && python3 -m pip install -e .
```
@ydshieh I do not understand why we don't:
1. Install from main on the docker build with `pip install git+https://github.com/hugginface/transformers@main` (and get all the necessary dependencies)
2. `pip uninstall -y transformers` in the docker 
3. `actions/checkout@v3`
4. `pip install -e  .`
(As discussed offline with @younesbelkada this should come in a separate PR)

ydshieh replied:
There is no deep reason for what we are doing currently: we just install `transformers` during docker image build and reuse it.

What you suggest (by installing `main` during docker image build) has a consequence that `we can't do experiment easily in a branch when we change `setup.py`).

Also by using `actions/checkout@v3`, the path to `transformers` will be different from the current approach, and therefore we have to change all the working directory path and artifact path, otherwise the report will be empty.

ArthurZucker reviewed (approved):
Preemptively approving, LGTM! Thanks @SunMarc for splitting the workflows!

ydshieh commented on .github/workflows/build-docker-images.yml:
```diff
@@ -262,38 +262,66 @@ jobs:
-  #       with:
-  #         context: ./docker/transformers-pytorch-deepspeed-amd-gpu
-  #         build-args: |
-  #           REF=main
-  #         push: true
-  #         tags: huggingface/transformers-pytorch-deepspeed-amd-gpu-push-ci
+    runs-on: [self-hosted, docker-gpu, amd-gpu, single-gpu, mi210]
+    steps:
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+      - name: Check out code
+        uses: actions/checkout@v3
+      - name: Login to DockerHub
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
+      - name: Build and push
+        uses: docker/build-push-action@v5
+        with:
+          context: ./docker/transformers-pytorch-deepspeed-amd-gpu
+          build-args: |
+            REF=main
+          push: true
+          tags: huggingface/transformers-pytorch-deepspeed-amd-gpu${{ inputs.image_postfix }}
+      # Push CI images still need to be re-built daily
+      -
+        name: Build and push (for Push CI) in a daily basis
+        # This condition allows `schedule` events, or `push` events that trigger this workflow NOT via `workflow_call`.
+        # The later case is useful for manual image building for debugging purpose. Use another tag in this case!
+        if: inputs.image_postfix != '-push-ci'
+        uses: docker/build-push-action@v5
+        with:
+          context: ./docker/transformers-pytorch-deepspeed-amd-gpu
+          build-args: |
+            REF=main
+          push: true
+          tags: huggingface/transformers-pytorch-deepspeed-amd-gpu-push-ci
```
This is irrelevant to this PR. We have (or at least had) issue of building AMD CI images and we haven't really make progress on it. 

Better to keep this as it is in this PR.

ydshieh reviewed (approved):
Thank you for this work! I could RIP with the daily CI now.

ydshieh commented:
I would love to see the newly added `job` could be run successfully (having failing tests are fine), but I don't enforce it here.

SunMarc committed:
revert changes

SunMarc committed:
Revert "revert changes"

This reverts commit 4cb52b8822da9d1786a821a33e867e4fcc00d8fd.

SunMarc committed:
revert correct change
